### ISSUE TITLE
"No MaterialLocalizations found" if explicitly specify locale on MaterialApp constructor

### ISSUE THREAD

**Issue by espresso3389** [flutter/flutter#22205]
On Android, to show texts with Japanese font correctly, we need to specify Japanese locale explicitly like the following sample:

```dart
class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return new MaterialApp(
      title: 'Flutter Demo',
      theme: new ThemeData(
        // This is the theme of your application.
        //
        // Try running your application with "flutter run". You'll see the
        // application has a blue toolbar. Then, without quitting the app, try
        // changing the primarySwatch below to Colors.green and then invoke
        // "hot reload" (press "r" in the console where you ran "flutter run",
        // or press Run > Flutter Hot Reload in IntelliJ). Notice that the
        // counter didn't reset back to zero; the application is not restarted.
        primarySwatch: Colors.blue,
      ),
      home: new MyHomePage(title: 'Flutter Demo Home Page'),
      locale: Locale('ja')
    );
  }
}
```

I think it's very common workaround for us as explained in several articles.
One of them is https://medium.com/@najeira/flutter-and-cjk-font-selection-2-2ba06e32dd8d

But the code above causes the following exception:

```
I/flutter (30658): ══╡ EXCEPTION CAUGHT BY WIDGETS LIBRARY ╞═══════════════════════════════════════════════════════════
I/flutter (30658): The following assertion was thrown building AppBar(dirty, state: _AppBarState#765c0):
I/flutter (30658): No MaterialLocalizations found.
I/flutter (30658): AppBar widgets require MaterialLocalizations to be provided by a Localizations widget ancestor.
I/flutter (30658): Localizations are used to generate many different messages, labels,and abbreviations which are used
I/flutter (30658): by the material library.
I/flutter (30658): To introduce a MaterialLocalizations, either use a  MaterialApp at the root of your application to
I/flutter (30658): include them automatically, or add a Localization widget with a MaterialLocalizations delegate.
```

I'm not sure whether it is expected behavior or not.
At least, certain old Flutter versions did not throw such exception.

My flutter environment is:

```
Running flutter doctor...
Doctor summary (to see all details, run flutter doctor -v):
[√] Flutter (Channel master, v0.9.3-pre.20, on Microsoft Windows [Version 10.0.18242.1000], locale ja-JP)
[√] Android toolchain - develop for Android devices (Android SDK 28.0.0)
[√] Android Studio (version 3.1)
[√] IntelliJ IDEA Ultimate Edition (version 2018.2)
[√] VS Code (version 1.27.1)
[√] Connected devices (1 available)
```

There's a very simple [repro (66.4 KB)](https://github.com/flutter/flutter/files/2410202/hello_world.zip), which is almost identical to flutter-create sample except two lines.

**Comment by endigo** [comment]
I faced the same error but this time my locale set to Mongolian.

Any updates?

**Comment by MrNickBarker** [comment]
After looking at `localizations.dart`, I see that the script variation `Latn` is handled as a country code. So in my case it should be `Locale('sr', 'Latn')` instead of `Locale('sr_Latn')` but unfortunately I'm still not able to support multiple variations (cyrillic and latinic scripts) since flutter just returns `sr_RS` regardless of device settings. My guess is that flutter is discarding script variations returned by the device.

**Comment by HansMuller** [comment]
@espresso3389 

> On Android, to show texts with Japanese font correctly, we need to specify Japanese locale explicitly ...

If the device is configured for Japanese, Flutter should follow suit. That said, to ensure that the app's locale is set correctly and that Material's localizations are available, you need to specify
the MaterialApp's `localizationsDelegates` and its `supportedLocales` list. Check out the[ Flutter internationalization tutorial](https://flutter.io/tutorials/internationalization/). 

An app that only supports Japanese (no matter how the device is configured) could be written like this:

```
import 'package:flutter_localizations/flutter_localizations.dart';

MaterialApp(
 localizationsDelegates: [
   GlobalMaterialLocalizations.delegate,
   GlobalWidgetsLocalizations.delegate,
 ],
 supportedLocales: [
    const Locale('ja', 'JP'),
 ],
  // ...
)
```

On iOS applications must also configure the `Info.plist` file. See ["Appendix: Updating the iOS app bundle"](https://flutter.io/tutorials/internationalization/#ios-specifics) in the internationalization tutorial.

**Comment by HansMuller** [comment]
@MrNickBarker 

You are correct, currently we only support Serbian locales 'sr' and `sr_Latn` .  If `supportedLocales` includes `Locale('sr')` Flutter will choose the Serbian translations so long as the device locale's language code is 'sr'.  In this case we are effectively discarding the locale's country code.

**Comment by HansMuller** [comment]
@espresso3389 That's a bit tricky, since the font selection is based on the locale. If you want and English app that renders strings using the Japanese font, you could try using `Localizations.override`, like the examples from https://github.com/flutter/flutter/pull/17879

```
new Localizations.override(
  context: context,
  locale: const Locale('ja'),
  child: new Text(character, style: style),
)
```

This way, the app can be localized for English (the default), but specific Text widgets can be localized for Japanese.

**Comment by espresso3389** [comment]
@HansMuller 
I understand it's a little off topic here but why does Flutter does not guess it's Locale from Android device's Locale? My devices are configured to use Japanese but Flutter seems to use English.

**Comment by HansMuller** [comment]
Flutter uses the device's locale as long as it is listed among the app's `supportedLocales`. If the device's locale is not found there, then `supportedLocales.first` is used. Typically that's english.

**Comment by espresso3389** [comment]
@HansMuller 
I see. So we should add `supportedLocales` or otherwise use `Localizations.override`.
Thanks for your kind explanation.

**Comment by HansMuller** [comment]
Fixed in #23850
